### PR TITLE
Feature socketcan bcm owned by bus

### PR DESCRIPTION
**Pull request by hardbyte**
Competes with #409  because I forgot to look if you were already doing this @christiansandberg !

* adds task management responsibility to the `Bus`. By default tasks created with `bus.send_periodic` will have a reference held by the bus - this means in many cases the user doesn't need to keep the task in scope for their periodic messages to continue being sent. If this behavior isn't desired pass `store_task=False` to the `send_periodic` method.
* the bus now has a `stop_all_periodic_tasks` method
* The socketcan tasks now reuse a bcm socket

### CODE REVIEW

File: can/interfaces/socketcan/socketcan.py [hunk 1]
```diff
@@ -563,7 +569,8 @@ def _send_once(self, data, channel=None):
     def send_periodic(self, msg, period, duration=None):
         """Start sending a message at a given period on this bus.
 
-        The kernel's broadcast manager will be used.
+        The kernel's broadcast manager will be used and the task will
+        be active while this Bus instance is still in scope.
```
Comment (christiansandberg) [inline can/interfaces/socketcan/socketcan.py]: ...or shutdown() or stop_all_periodic_tasks() is called?

**Comment by felixdivo** [comment]
Note: We might also consider calling `shutdown()` in the [`__del__`](https://docs.python.org/2/reference/datamodel.html#object.__del__) method of the bus.

**Comment by hardbyte** [comment]
Hmm okay what about this slightly hack approach of wrapping the `task.stop()` method inside `bus.send_periodic`. There would still be issues with resumable tasks and expired tasks though.

**Comment by hardbyte** [comment]
Regarding the memory leak issue at the least we should clearly document what the user can do to prevent it, but I'd rather have an api that allows the user to take ownership for the task management if they want to. Perhaps a flag in the bus initializer or the `send_periodic` method?

File: can/bus.py [hunk 2]
```diff
@@ -303,6 +339,9 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.shutdown()
 
+    def __del__(self):
+        self.shutdown()
```
Comment (christiansandberg) [inline can/bus.py]: Not all implementations may be able to handle shutdown being called twice.

Comment (hardbyte) [reply can/bus.py]: Fair point - I'd rather remove the `__del__` from the bus instead of adding shutdown state tracking. Do you agree?

**Comment by hardbyte** [comment]
I think this is pretty much ready - I've just rebased

File: can/bus.py (hunk 2 above)
Comment (christiansandberg) [reply can/bus.py]: I think so too.

**Comment by felixdivo** [comment]
I just merged this. Someone else (@christiansandberg ?) needs to have a look at #292 and #404.

**Pull request merged**